### ISSUE TITLE
TST: We should add missing tests for cdflib

### ISSUE THREAD

**Issue by steppi** [scipy/scipy#19953]
Since merging https://github.com/scipy/scipy/pull/19560,  @ilayn's translation of cdflib into cython, several bugs have been brought to our attention by statsmodels maintainer @bashtage, https://github.com/scipy/scipy/issues/19896, https://github.com/scipy/scipy/pull/19933, https://github.com/scipy/scipy/pull/19944, who kindly made PRs fixing several of them.

Currently `scipy/special/tests/test_cdflib.py` is missing tests for a number of functions

https://github.com/scipy/scipy/blob/11ce9d798116092189a25d3799a6bcecfdedbc53/scipy/special/tests/test_cdflib.py#L4-L19

I think there are two primary types of tests we can add. Tests against mpmath, (which will require writing arbitrary precision implementations of these functions using mpmath), and tests against the Fortran version (which would be faster, but it won't catch bugs in the Fortran version).

@ilayn offered to add tests if I came up the parameter domains which in which the functions should be tested. I started making such a list, and realized that such special care around hot spots probability isn't necessary just yet. I think we should just use the classes `Arg`, `IntArg`, and `ProbArg` used in `test_cdflib.py` over the domains of definition of the functions. When mpmath is used, we should restrict the domains for cases where mpmath cannot compute a correct answer for sufficiently large values.

While going through all of these, I found that the functions for the non-Central F distribution, `ncfdtr` and friends are documented to require the non-centrality parameter `nc` to in $(0, 1e^4)$. The comment in `_cdflib.pyx` refers to the bound too. I've found the bound is missing from the code though. @ilayn, do you remember if there was a reason to take the bound out? I can't remember if that came up in the review. We'll either need to put the bound back or change the docs.

Bounds below.

### Non-central F distribution
* `ncfdtr`, `ncfdtri`, `ncfdtridfn`, `ncfdtridfd`, and `ncfdtrinc`.
    * `dfn`: `Arg(0, np.inf, inclusive_a=False)`
    * `dfd`: `Arg(0, np.inf, inclusive_a=False)`
    * `nc`: `Arg(0, 1e4)` This is the documented range at least, but it looks like this check has been removed.
    * `f`: `Arg(0, np.inf)`
    * `p`: `ProbArg()`.

### Negative binomial distribution
* `nbdtrik` and `nbdtrin`
    * `k`: `IntArg(1, 1000)`
    * `y`: `ProbArg()`
    * `n`: `IntArg(1, 1000)`
    * `p`: `ProbArg()`

### Normal distribution
* `nrdtrimn` and `nrdtrimsd`
    * `p`: `ProbArg()`
    * `x`: `Arg(-np.inf, np.inf)`
    * `std`: `Arg(0, np.inf, inclusive_a=False)`
    * `mn`: `arg(-np.inf, np.inf, inclusive_a=False, inclusive_b=False)`

### Poisson distribution
* `pdtrik`
    * `p`: `ProbArg()`
    * `m`: `Arg(0, np.inf, inclusive_b=False)`

### Non-central T distribution
* `ncdtr`, `nctdtrit`, `nctdtridf`, and `nctdrinc`.
    * `df`: Arg(0, np.inf, inclusive_a=False)
    * `nc`: `Arg(-1e6, 1e6)`
    * `t`: `Arg(-np.inf, np.inf)`
    * `p`: `ProbArg()`

**Comment by ilayn** [comment]
If there is such a bound then it should be in the wrappers as fortran code does not check it 

This is the old wrapper 

https://github.com/scipy/scipy/blob/1d3a067c2ccd0a6efddeb3194163aa9a3879d26e/scipy/special/cdf_wrappers.c#L243-L252

This is the fortran code where it uses the 10^4 because it is an inverse search 

https://github.com/scipy/scipy/blob/1d3a067c2ccd0a6efddeb3194163aa9a3879d26e/scipy/special/cdflib/cdffnc.f#L261-L285

So it is only the `which=5` that uses that bound which makes sense because it is a bound of search failure. And that failure is in the code

https://github.com/scipy/scipy/blob/11ce9d798116092189a25d3799a6bcecfdedbc53/scipy/special/_cdflib.pyx#L2539

Hence that's not missing anywhere.

**Comment by steppi** [comment]
Got it. I think it's a docs problem. The docs mention the bound for all of the non-central f functions, making it seem like that should be the accepted domain.

![image](https://github.com/scipy/scipy/assets/1953382/eaeabddd-9f28-45d1-bf19-a10be3a4b145)

**Comment by ilayn** [comment]
Yeah, this bound should only appear on `ncfdtrinc` as an algorithmic shortcoming but it doesn't https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ncfdtrinc.html#scipy.special.ncfdtrinc

**Comment by dschmitz89** [comment]
I would like to help out with this issue.

I started looking into the normal distribution related functions `nrdtrimn` and `nrdtrisd`. I am probably misunderstanding the function signature as for me a very basic test already fails, returning `nan` for `nrdtrimn`:

```python
from scipy.stats import norm
from scipy.special import nrdtrimn

x, mu, sigma = 0, 0, 0.1
q = norm.cdf(x, loc=mu, scale=sigma)  # 0.5

nrdtrimn(q, x, sigma) -> NaN
``` 

Am I missing something?

**Comment by ilayn** [comment]
Yes this is a very funky way of wrapping things for me too. So typically what I do is first get the function name to be tested, from the test file. Then I search for it in the `function.json` that gives me the wrapper name in the `cdflib_wrappers` to search for. Then I try to mimic the input to the output to test the function. It is indeed very cumbersome process overall probably for historical reasons. The naming scheme of these functions are very confusing for me.

**Comment by dschmitz89** [comment]
Thanks for the hint. This solved the mystery to me: the docstring(https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.nrdtrimn.html#scipy.special.nrdtrimn) is currently wrong, it lists the arguments in the wrong order. It must be `nrdtrimn(p, std, x)` instead of  `nrdtrimn(p, x, std)` . Will send a fix for that and a test soon.

**Comment by ilayn** [comment]
OK in #20069 we tried 

```python
def _normal_cdf(sigma, x, mu):
    x = mpmath.mp.mpf(x)
    mu = mpmath.mp.mpf(mu)
    sigma = mpmath.mp.mpf(sigma)
    q = mpmath.mp.ncdf(x, mu, sigma)
    return q

[snipped]
class
    def test_nrdtrimn(self):
        _assert_inverts(
            sp.nrdtrimn,
            _normal_cdf,
            1, [ProbArg(), Arg(-30, 30), Arg(0, 10, inclusive_a=False)],
            2, [ProbArg(), Arg(0, np.inf, inclusive_a=False),
                Arg(-np.inf, np.inf)],
            n=1000, rtol=1e-5)
```

Somehow couldn't make it work because the comparison was being done on the wrong argument. If we have an understanding why then we can roll out most of these in a single PR.

**Comment by ilayn** [comment]
I managed to make this work but the main issue is the closed form mpmath or some other credible source to compare for the other ones. For now, I'm just comparing the CDF with the inverse functions at least to stay consistent.

**Comment by steppi** [comment]
For the `mpmath` baselines, I think the best thing to do is use mpmath's [findroot](https://mpmath.org/doc/current/calculus/optimization.html#root-finding-findroot)  to invert an mpmath implementation of the cdf function and retry with a higher `dps` if the tolerance isn't met.

**Comment by dschmitz89** [comment]
Can some `ncfdtr` related ones be ticked off after https://github.com/scipy/scipy/pull/21505 @steppi ? Feel free to ignore if updating this testing framework is anyway part of your cdflib rework.

**Comment by JacobHass8** [comment]
Has much of this been addressed in #21966? I would like to work on adding more test coverage but I don't want to repeat work that has already been done.

**Comment by dschmitz89** [comment]
`nbdtrik` is done in #24711 but the others are still open.

Ideally we could just reuse the existing spot tests for the inverses as well like I did in #24711 .

**Comment by JacobHass8** [comment]
> [@ilayn](https://github.com/ilayn) offered to add tests if I came up the parameter domains which in which the functions should be tested. I started making such a list, and realized that such special care around hot spots probability isn't necessary just yet. I think we should just use the classes `Arg`, `IntArg`, and `ProbArg` used in `test_cdflib.py` over the domains of definition of the functions. When mpmath is used, we should restrict the domains for cases where mpmath cannot compute a correct answer for sufficiently large values.

Is the correct way to do this still with `_assert_inverts`? I've been trying to get a simple test running but haven't had any luck. I vaguely understand what the inputs should be but I'm getting errors that I can't parse. For example, here is the simplest test I could think of 
```python
def test_nbdtrik(self):
    _assert_inverts(
        sp.nbdtrik,
        _nbdtr,
        0, [IntArg(1,10), IntArg(1, 10), Arg(0.1, 0.9)], 
        spfunc_first=False,        
        rtol=1e-4, 
        n=10)
```
where 
```python
def _nbdtr(k, n, p):
  # Throws a warning if k is not an integer
  with warnings.catch_warnings(action="ignore"):
      return mpmath.mpf(sp.nbdtr(k, n, p))
```
My understanding is that the arguments to `_assert_inverts` is as follows. The first argument is the function doing the inverting, in this case `nbdtrik`. The second argument returns the cdf of the probability distribution (this should be written in mpmath to be as accurate as possible. I didn't do that here just to be quick). The third argument is the index of the argument in the second function that we are inverting with respect to. Since we are inverting with respect to `k` and `_nbdtr(k,n,p)` takes `k` as the first parameter this will be 0. I don't understand the remaining parameters though. The fourth argument is a list of the parameter space for the second function since `spfunc_first=False`?  Why can't I just input the arguments to the first function, `sp.nbdtrik`, and take spfunc_first=True? I did this previously and only got a ton of errors. 

~Anyways, I'll keep trucking along and see how things go.~ Nevermind, I don't have a clue how to use `_assert_inverts`. I'm getting values of `nan` for parameters that I've tested and should be correct. I could dig in more, but I there's no documentation for the class so I'm going to put it off.

**Comment by dschmitz89** [comment]
I could not get that machinery to work either. I see a few major issues with it:
* mpmath rootfinding can take practically forever if you need to set the number of significant sigits very high (which we need for noncentral T and noncentral F for example)
* mpmath CDF computations not being able to converge (observed for incomplete beta functions which you need for `nbdtrik`)
* Little control over which points are tested. This will lead to issues where the machinery repeatedly tests failing edge cases

All in all, I would advise against using this and just rely on spot checks. That's what we did for all functions here except `nrdtrimn` and `nrdtrisd`. Note that some of the tests in this machinery were also xfailed in the past (for example [bdtrik](https://github.com/scipy/scipy/blob/ae5a06b07593de27c42e9da88044b6fcc39d6172/scipy/special/tests/test_cdflib.py#L204)).

**Comment by JacobHass8** [comment]
Ugh, it seems so close but maybe it's not meant to be. I had really hoped there would be a more general method available tham just spot checking.

**Comment by ilayn** [comment]
At this point, I think it is obvious that none of us managed to take advantage of that heavily layered indirection testing infra.

So if there is time and effort to be wasted on these tests, I suggest getting mpmath values offline and write regular tests that are flattened and do test values directly. There is no point trying to squeeze into that framework. 

If inversion or round trip is to be tested, it can still be done per function instead and would give immediate point where it failed and with which values. It was a pain in the ass to find which value failed in the current infra.

**Comment by JacobHass8** [comment]
Great, sorry for bugging everyone about this. I'll write up some generic spot checks using mpmath to calculate the cdf and open a PR. Thanks for the insight!

**Comment by dschmitz89** [comment]
@JacobHass8 Just a heads up: the list for #21966 is currently done until the next boost release. Only then we can use the new noncentral T and noncentral F parameter finders. I opened PRs #24695 and #24711 in case you would like to help with reviews.
